### PR TITLE
Add 'Fingerprint' Currency Type... And Serialize It!

### PR DESCRIPTION
**Pull request by CodaFi**
* Add a new `Fingerprint` currency type that centralizes a bunch of hashing invariants and gives me some room to try to experiment with different hashing regimes.
* Push the `Fingerprint` type as a refinement of the `std::string`-ified interface we had there previously.

There's lots of plumbing here. The most fiddly bits are in the new serialization code. We'll be serializing fingerprints in a side table for swiftmodules. This ensures that fingerprints are preserved for cross-module dependencies so we can properly take advantage of type body fingerprints.

Finally, migrate the "fingerprints" in some hand-written tests. These tests were using fingerprints name 'after', 'before' and 'same', which are far too short to resemble a fingerprint value, which must be 32-bytes. Migrate these files so that

'after' -> 'afterachangewasmadetoswiftsource'
'before' -> 'beforechangewasmadetoswiftsource'
'same' -> 'sameasiteverwassameasiteverwassa'

There is otherwise no change to the contents of these files.

The driver's unit tests are getting a similar treatment, since it uses a user-provided name of the swiftdeps file for the "job" in the request graph. The unit test driver will now extend those names to 32-bytes by padding them with `X`s.

rdar://71343649

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test source compatibility

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test

File: lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -547,10 +547,7 @@ populateOutOfDateMap(InputInfoMap &map, llvm::sys::TimePoint<> &LastBuildTime,
       // -swift-version argument is handled in the argsHashStr check that
       // follows.
       CompilationRecordSwiftVersion = value->getValue(scratch);
-      versionValid = (CompilationRecordSwiftVersion
-                      == version::getSwiftFullVersion(
-                        version::Version::getCurrentLanguageVersion()));
-
+      versionValid = CompilationRecordSwiftVersion.equals(version::getSwiftFullVersion(version::Version::getCurrentLanguageVersion()));
```
Comment (CodaFi) [inline lib/Driver/Driver.cpp:550]: N.B. This is a mistake. I will revert this after the tests pass.

**Review by davidungar — CHANGES_REQUESTED** [review]
This is great stuff, and I want to just approve it. But, from my reading of the comment for the `Fingerprint` type, you and I disagree on whether a fingerprint reflects identity. Since I believe that a fingerprint reflects state, and that it is useful to think of a declaration has having an identity independent of its state (e.g. adding a method to a struct does not change the struct's identity), I don't think that the comment as-is is accurate.

But, maybe I'm wrong, or there is a better framing I don't see. Perhaps it would help to talk offline?

**Review by davidungar — COMMENTED** [review]
Spoke offline, there will be a wording change to the one comment, and a change to the "000..." part.

**Review by davidungar — APPROVED** [review]
LGTM!

File: include/swift/Basic/Fingerprint.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+namespace llvm {
+namespace yaml {
+class IO;
+}
+}; // namespace llvm
+
+namespace swift {
+
+/// A \c Fingerprint represents a stable point of identity for a piece of data
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply include/swift/Basic/Fingerprint.h]: Spoke about this offline, we chose to use `stable summary of the contents...`

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
⛵

**Pull request merged**